### PR TITLE
Prevent submission of payment requests that have no documentation

### PR DESCRIPTION
**Pull request by coreymckrill**
For both Payment Requests and Reimbursement Requests:

* Add a method to check and see if the request post can be submitted. Currently this just checks to see if the post has one or more files attached to it.
* When the post fails the check, show a note in the Status metabox that documentation must be attached first, and don't render the _Submit for Review_ button.

Screenshots:

Payment request with no attachments:
payment-request_no-attachment

Reimbursement request with no attachments:
reimbursement-request_no-attachment

### CODE REVIEW

**Review by vedanshujain — APPROVED** [review]
LGTM

**Comment by hlashbrooke** [comment]
Looks good.

File: public_html/wp-content/plugins/wordcamp-payments/includes/payment-request.php [hunk 1]
```diff
@@ -230,16 +239,21 @@ public function render_status_metabox( $post ) {
 			$post->post_status = $back_compat_statuses[ $post->post_status ];
 		}
 
-		$editable_statuses = array( 'auto-draft', 'draft', 'wcb-incomplete' );
+		$editable_statuses = self::get_editable_statuses();
```
Comment (iandunn) [inline public_html/wp-content/plugins/wordcamp-payments/includes/payment-request.php]: Good job making this DRY.

File: public_html/wp-content/plugins/wordcamp-payments/includes/payment-request.php [hunk 2]
```diff
@@ -454,6 +468,21 @@ protected function render_files_input( $post, $label, $name, $description = '' )
...
+	 * @param WP_Post $post
+	 */
+	protected function can_submit_request( $post ) {
+		$files = WordCamp_Budgets::get_attached_files( $post );
+
+		if ( empty( $files ) ) {
+			return false;
+		}
+
+		return true;
```
Comment (iandunn) [inline public_html/wp-content/plugins/wordcamp-payments/includes/payment-request.php]: Could this be simplified to `return ! empty( $files );` ?

File: public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php [hunk 3]
```diff
@@ -289,8 +291,13 @@ function render_status_metabox( $post ) {
...
 	} elseif ( in_array( $post->post_status, $editable_statuses ) ) {
-		$submit_text = esc_html__( 'Submit for Review', 'wordcamporg' );
-		$submit_note = esc_html__( 'Once submitted for review, this request can not be edited.', 'wordcamporg' );
+		if ( can_submit_request( $post ) ) {
+			$submit_text = esc_html__( 'Submit for Review', 'wordcamporg' );
+			$submit_note = esc_html__( 'Once submitted for review, this request can not be edited.', 'wordcamporg' );
+		} else {
+			$submit_note = esc_html__( 'Documentation must be attached and saved to the draft before this request can be submitted.', 'wordcamporg' );
+		}
+
```
Comment (iandunn) [inline public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php]: Do we want to make this DRY also, since it seems identical?

File: public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php [hunk 4]
```diff
@@ -404,6 +410,21 @@ function display_post_states( $states ) {
...
+function can_submit_request( $post ) {
+	$files = WordCamp_Budgets::get_attached_files( $post );
+
+	if ( empty( $files ) ) {
+		return false;
+	}
+
+	return true;
+}
+
```
Comment (iandunn) [inline public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php]: seems like this could also be DRY

File: public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php [hunk 5]
```diff
@@ -250,7 +252,7 @@ function enqueue_assets() {
  * requests and to submit notes, but they should only be able to change the form fields if the post hasn't been
  * submitted yet, or if we've asked for more information.
  *
- * @param \WP_Post $post
+ * @param WP_Post $post
```
Comment (iandunn) [inline public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php]: Since there's a lot of PHPCS fixes, I think it'd be ideal for them to be in a separate commit, to avoid diff noise, but it's probably not worth breaking the commits apart now that they're already pushed.

File: public_html/wp-content/plugins/wordcamp-payments/views/payment-request/metabox-status.php [hunk 6]
```diff
@@ -1,3 +1,13 @@
+<?php
+/** @var WP_Post $post */
+/** @var WCP_Payment_Request $this */
+/** @var bool $current_user_can_edit_request */
+/** @var string $submit_text */
+/** @var string $submit_note */
+/** @var bool $date_vendor_paid_readonly */
+/** @var string $incomplete_notes */
+/** @var bool $incomplete_readonly */
+?>
```
Comment (iandunn) [inline public_html/wp-content/plugins/wordcamp-payments/views/payment-request/metabox-status.php]: This would be a bit more readable if the `$` were aligned.

File: public_html/wp-content/plugins/wordcamp-payments/views/payment-request/metabox-status.php [hunk 7]
```diff
@@ -86,23 +96,24 @@ class="large-text"
...
-				<?php if ( current_user_can( 'delete_post', $post->ID ) ) : ?>
+			<?php if ( current_user_can( 'delete_post', $post->ID ) ) : ?>
+				<div id="delete-action">
 					<a class="submitdelete deletion" href="<?php echo get_delete_post_link( $post->ID ); ?>">
 						<?php _e( 'Delete', 'wordcamporg' ); ?>
 					</a>
-				<?php endif; ?>
-			</div>
+				</div>
+			<?php endif; ?>
```
Comment (iandunn) [inline public_html/wp-content/plugins/wordcamp-payments/views/payment-request/metabox-status.php]: I wonder if the `delete-action` element was intentionally left in, even if it's blank, maybe to preserve the layout or something? It might be affected by cases where `#publishing-action` is shown/not-shown.

If it tests well for you though, then it's probably safe to change this. It does make more sense to not create it when the button won't be created, and we can always revert this part if we find a problem in the future.

**Review by iandunn — APPROVED** [review]
Left some minor thoughts but overall LGTM

**Comment by sarahmonster** [comment]
Some thoughts here from a user experience perspective:

- Change the wording of the notice here to a more active voice—something along the lines of "Please attach an invoice, contracts, or other vendor correspondence to submit your payment request. Your request will be saved as a draft until files are attached."
- Highlight the files field to indicate where this documentation should be attached.
- Use a standard error notice styling and placement.
- It may make sense to nix the notes field and the bit about "If no supporting documentation is available, please indicate the reason in the notes below." as well, unless submitting a request without an attachment, but with text in this field, would be allowed through.

I can put a quick mockup of this together if that's helpful!

File: public_html/wp-content/plugins/wordcamp-payments/views/payment-request/metabox-status.php (hunk 7 above)
Comment (coreymckrill) [reply public_html/wp-content/plugins/wordcamp-payments/views/payment-request/metabox-status.php]: I haven't noticed any issues in testing. I'd prefer to leave it until someone reports a problem.

Comment (iandunn) [reply public_html/wp-content/plugins/wordcamp-payments/views/payment-request/metabox-status.php]: 👍

File: public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php (hunk 3 above)
Comment (iandunn) [reply public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php]: `wordcamp-payments/wordcamp-budgets.php` has some shared stuff already, so that might be good.

File: public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php (hunk 5 above)
Comment (coreymckrill) [reply public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php]: Good point

File: public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php [hunk 8]
```diff
@@ -287,15 +287,17 @@ function render_status_metabox( $post ) {
...
 		$current_user_can_edit_request = true;
 	} elseif ( in_array( $post->post_status, $editable_statuses ) ) {
-		if ( can_submit_request( $post ) ) {
+		if ( WordCamp_Budgets::can_submit_request( $post ) ) {
 			$submit_text = esc_html__( 'Submit for Review', 'wordcamporg' );
-			$submit_note = esc_html__( 'Once submitted for review, this request can not be edited.', 'wordcamporg' );
+			$submit_note = esc_html__( 'Once submitted for review, this request cannot be edited.', 'wordcamporg' );
 		} else {
-			$submit_note = esc_html__( 'Documentation must be attached and saved to the draft before this request can be submitted.', 'wordcamporg' );
+			$submit_note = esc_html__( 'Please add an invoice or other supporting documentation in the Files section and save the draft.', 'wordcamporg' );
```
Comment (iandunn) [inline public_html/wp-content/plugins/wordcamp-payments/includes/reimbursement-request.php]: Should this be escaped when it's output rather than here?

**Comment by coreymckrill** [comment]
Committed to SVN: https://meta.trac.wordpress.org/changeset/8136
